components/style/values/generics/rect.rs

```diff
@@ -0,0 +1,110 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Generic types for CSS values that are composed of four sides.
+
+use cssparser::Parser;
+use parser::{Parse, ParserContext};
+use std::fmt;
+use style_traits::ToCss;
+
+/// A CSS value made of four sides: top, right, bottom, and left.
+#[derive(Clone, Copy, Debug, HasViewportPercentage, PartialEq, ToComputedValue)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct Rect<T> {
+    /// Top
+    pub top: T,
+    /// Right.
+    pub right: T,
+    /// Bottom.
+    pub bottom: T,
+    /// Left.
+    pub left: T,
+}
+
+impl<T> Rect<T> {
+    /// Returns a new `Rect<T>` value.
+    pub fn new(top: T, right: T, bottom: T, left: T) -> Self {
+        Rect {
+            top: top,
+            right: right,
+            bottom: bottom,
+            left: left,
+        }
+    }
+}
+
+impl<T> Rect<T>
+    where T: Clone
+{
+    /// Parses a new `Rect<T>` value with the given parse function.
+    pub fn parse_with<Parse>(
+        context: &ParserContext,
+        input: &mut Parser,
+        parse: Parse)
+        -> Result<Self, ()>
+        where Parse: Fn(&ParserContext, &mut Parser) -> Result<T, ()>
+    {
+        let top = parse(context, input)?;
+        let right = if let Ok(right) = input.try(|i| parse(context, i)) { right } else {
```

Comment:
I tried it, this ended up being more verbose.

Best solution would be https://github.com/rust-lang/rfcs/pull/1303#issuecomment-303396592.